lib/datadog/core/configuration/option.rb

```diff
@@ -94,6 +96,87 @@ def default_precedence?
 
         private
 
+        def coerce_env_variable(value)
+          case @definition.type
+          when :int
+            value.to_i
+          when :float
+            value.to_f
+          when :array
+            values = if value.include?(',')
+                       value.split(',')
+                     else
+                       value.split(' ') # rubocop:disable Style/RedundantArgument
+                     end
+
+            values.map! do |v|
+              v.gsub!(/\A[\s,]*|[\s,]*\Z/, '')
+
+              v.empty? ? nil : v
+            end
+
+            values.compact!
+            values
+          when :bool
+            string_value = value
+            string_value = string_value.downcase
+            string_value == 'true' || string_value == '1' # rubocop:disable Style/MultipleComparison
+          else
+            value
+          end
+        end
+
+        def validate_type(value)
+          raise_error = false
+
+          valid_type = validate(@definition.type, value)
+
+          unless valid_type
+            raise_error = if @definition.type_options[:nil]
+                            !value.is_a?(NilClass)
+                          else
+                            true
+                          end
+          end
+
+          if raise_error
+            error_msg = if @definition.type_options[:nil]
+                          "The option #{@definition.name} support this type `#{@definition.type}` "\
+                                      "and `nil` but the value provided is #{value.class}"
+                        else
+                          "The option #{@definition.name} support this type `#{@definition.type}` "\
+                          "but the value provided is #{value.class}"
+                        end
+
+            raise ArgumentError, error_msg
+          end
+
+          value
+        end
+
+        def validate(type, value)
+          case type
+          when :string
+            value.is_a?(String)
+          when :int
+            value.is_a?(Integer)
+          when :float
+            value.is_a?(Float)
+          when :array
+            value.is_a?(Array)
+          when :hash
+            value.is_a?(Hash)
+          when :bool
+            value.is_a?(TrueClass) || value.is_a?(FalseClass)
+          when :block
+            value.is_a?(Proc)
+          when :symbol
+            value.is_a?(Symbol)
+          else
+            true
```

Comment:
I think we should emit a message through `Datadog.logger` here to let the user know that their type wasn't recognized.

Reply:
I see two cases here:

- an option type is not enforced, `type` is presumably `nil` and `validate` should return `true` because all value types are accepted:

  ```
  when nil
    true
  ```

  Is that something we want, or should we disallow unenforced types altogether?
  
- an option definition receives an unknown type, e.g:
  
  ```
  option :foo do |o|
    o.type :frobz
  end
  ```
  
  Then the type cannot be validated, but that error is _on us_, because the option definition is incorrect, so to me `validate` should blow up:
  
  ```
  else
    raise ConfigurationSettingError, "unknown type: #{type}"
  ```

Reply:
Right now, when the `type` is `nil` it is a way to signal that we do not want any validation or coercion.

> Is that something we want, or should we disallow unenforced types altogether?

That would be the goal, unfortunately, the current configuration options include some complex settings that coercion and validation differ. An example is the `tags` coerce the environment variable into an array and the supported value for setting the option via the configure block is a hash. 

A possible solution would be to add some functionality to the coercion to be able to get the string value to a hash. 

As for invalid type coercion of validation I like the idea to raise an exception ex: `ConfigurationSettingError `

Reply:
Added the code to raise an exception if the type value is invalid here: https://github.com/DataDog/dd-trace-rb/pull/2970/commits/7a476a00f0bf7c728f80c34660507e61c8891d97

Reply:
> when the `type` is `nil` it is a way to signal that we do not want any validation or coercion.

I realised that when seeing the other discussion about delegating complex type checking to the setter, but thanks for the clarification.